File: src/hritserver/mime/Part.java
/* This file is part of hritserver.
 *
 *  hritserver is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  hritserver is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with hritserver.  If not, see <http://www.gnu.org/licenses/>.
 */

package hritserver.mime;
import java.io.ByteArrayOutputStream;
import hritserver.exception.HritException;
/**
 *
 * @author desmond
 */
public abstract class Part 
{
    ByteArrayOutputStream body;
    String encoding;
    Header header;
    static String CRLF = "\r\n";
    public Part( String encoding )
    {
        this.encoding = encoding;
        header = new Header();
        body = new ByteArrayOutputStream();
    }
    /**
     * Convert the part to a string
     * @return an empty string if the encoding didn't work
     */
    public String toString()
    {
        String ret = "";
        try
        {
            ret = header.toString()+body.toString( encoding );
        }
        catch ( Exception e )
        {
            System.err.println(e.getMessage());
        }
        return ret;
    }
    /**
     * Get the length of this part
     * @return an integer being the overall length of this part
     */
    int getLength()
    {
        return header.length()+body.size();
    }
    /**
     * Add a header to the part
     * @param header 
     */
    void addHeader( String header )
    {
        this.header.add( header );
    }
    /**
     * Retrieve the header
     * @return a Header
     */
    public Header getHeader()
    {
        return header;
    }
    /**
     * Get the body as a string
     * @return a string being only the body not the headers
     */
    public String getBody()
    {
        return this.body.toString();
    }
    /**
     * Add some text to the body of this part
     * @param text the text to add
     */
    void addToBody( byte[] body )
    {
        this.body.write( body, 0, body.length );
    }
    /**
     * Turn a string into a part
     * @param text the text to convert
     * @return a subclass of Part
     */
    static Part parse( String text ) throws HritException
    {
        Part p = null;
        String sep = CRLF+CRLF;
        int bodyPos = text.indexOf(sep);
        if ( bodyPos == -1 )
            throw new HritException("no division between header and body");
        else
        {
            Header h = Header.parse( text.substring(0,bodyPos) );
            p = h.createPart();
            String body = text.substring( bodyPos+sep.length() );
            try
            {
                p.addToBody( body.getBytes(h.getEncoding()) );
            }
            catch ( Exception e )
            {
                throw new HritException( e );
            }
        }
        return p;
    }
}

File: calliope-0.2.2/pdef-tool/src/part.c
/*
 * This file is part of mmpupload.

    mmpupload is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    mmpupload is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with mmpupload.  If not, see <http://www.gnu.org/licenses/>.*/
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "part.h"
#ifdef MEMWATCH
#include "memwatch.h"
#endif
/*
 * Represent part of a mime multipart upload
 */
struct part_struct
{
    char *name;
    char *encoding;
    unsigned char *body;
    part_kind kind;
    int len;
    struct part_struct *next;
};
/**
 * Create a part for later writing out
 * @param name the name of the part which may be a form value
 * @param value the form value or body of file already allocated
 * @param len the length of the body
 * @param kind the kind of the part: form data or file contents
 * @param enc the mime type of the data
 */
part *part_create( const char *name, unsigned char *value, int len, 
    part_kind kind, const char *enc )
{
    part *p = calloc( 1, sizeof(part) );
    if ( p != NULL )
    {
        p->name = strdup( name );
        p->encoding = strdup( enc );
        p->body = value;
        p->kind = kind;
        p->len = len;
        if ( p->name == NULL || p->encoding == NULL )
        {
            fprintf(stderr,"part: failed to copy name or encoding\n");
            free( p );
            p = NULL;
        }
    }
    else
        fprintf(stderr,"part: failed to allocate object\n");
    return p;
}
/**
 * Dispose of a part
 * @param p the part to delete
 */
void part_dispose( part *p )
{
    if ( p->name != NULL )
        free( p->name );
    if ( p->encoding != NULL )
        free( p->encoding );
    if ( p->body != NULL )
        free( p->body );
    free( p );
}
/**
 * Compute the exact size of the part
 * @param p the part in question
 * @param boundary the boundary to use
 * @return the exact amount of memory required by the part
 */
int part_size( part *p, char *boundary )
{
    int used = 0;
    if ( p->kind == formdata )
    {
        used = strlen(boundary)+4;
        used += 43+strlen(p->name);
        used += 2+strlen((char*)p->body);
    }
    else // file
    {
        used = 4+strlen( boundary );
        used += 68+strlen( p->name );
        used += 18+strlen( p->encoding );
        //if ( p->len > 0 )
        //    used += 35; //  content transfer encoding
        used += p->len + 2;
    }
    return used;
}
/**
 * Get the next part in the list
 * @param p the part preceding it
 * @return the next part
 */
part *part_next( part *p )
{
    return p->next;
}
/**
 * Attach this part to the next
 * @param q the part to attach to
 * @param p the part to attach to it
 */
void part_set_next( part *q, part *p )
{
    q->next = p;
}
/**
 * Write the part into the destination data
 * @param p the part to write
 * @param body the body of data to write into
 * @param boundary the mmp boundary to use
 * @param len the length consumed in body
 */
int part_get( part *p, unsigned char *body, char *boundary, int len )
{
    int used = 0;
    if ( p->kind == formdata )
    {
        used = snprintf( (char*)body, len, "--%s\r\n",boundary );
        used += snprintf( (char*)&body[used], len-used,
            "Content-Disposition: form-data; name=\"%s\"\r\n\r\n", p->name );
        used += snprintf( (char*)&body[used], len-used, "%s\r\n", 
            (char*)p->body );
    }
    else // file
    {
        used = snprintf( (char*)body, len, "--%s\r\n",boundary );
        used += snprintf( (char*)&body[used], len-used,
            "Content-Disposition: form-data; name=\"uploadedfile[]\"; "
            "filename=\"%s\"\r\n", p->name );
        used += snprintf( (char*)&body[used], len-used, 
            "Content-Type: %s\r\n\r\n", (char*)p->encoding );
        if ( p->len > 0 )
        {
            memcpy( &body[used], p->body, p->len );
            used += p->len;
        }
        memcpy ( &body[used], "\r\n", 2 );
        used += 2;
    }
    return used;
}
